docs/DATASETS_v3.md

```diff
@@ -0,0 +1,355 @@
+
+`[77]` -> MessagePack -> `bytes([0x91, 0x4d])` -> Base64 -> `kU0=`
+
+#### Path to the feature file
+
+For technical reasons, it is best if only a relatively small number of features are stored together in a single directory, and similarly if only a small number of directories are stored together in a single directory. Ideally, the features created at the same time or likely to be edited at the same time should be stored together, rather than spread out among all the other features - so, neighbouring primary key values should be neighbouring files where possible.
+
+The exact system used to generate the path to the file depends on a few parameters which are stored in the dataset as an extra meta item called `path-structure.json`. The path structure might look like this:
+
+```json
+{
+  "scheme": "int",
+  "branches": 64,
+  "levels": 4,
+  "encoding": "base64"
+}
+```
+
+The `"scheme": "int"` tells us that this path-structure is used for a dataset which has a single primary key column of type integer, and that value will be used directly to generate the path to the file. (The only other supported scheme is `"msgpack/hash"` - see below).
+
+The next two parameters - `"branches": 64, "levels": 4` indicate that there are 4 levels of directory hierarchy, and at each level, there are up to 64 different directories branching out, such that a dataset with a huge number of features will have them spread across `64 ** 4 = 16777216` leaf-node directories - so a dataset could have `64 ** 5 = 1073741824` features and no directory would contain more than 64 directories or features. (Directories are only created when needed, so a dataset with only one feature with primary key 1 would create only four nested folders in which to store it, eg `A/A/A/A`.)
+
+Each directory is named after a character in the base64 alphabet - this is the `"encoding": "base64"`, and this encoding only supports a branch factor of 64. The other valid encoding is `"hex"`, which supports a branch factor of 16 or 256.
+
+So to encode the example before where the primary-key-value-array is `[77]` - since the scheme is "int" we know there is only one primary key value, an integer, which we can use as input for the subsequent steps: `77`. Encoding an integer (rather than a string of bytes) using Base64 works similarly to encoding integers in other bases such as hexadecimal. A quick primer: 0 is `A`, 1 is `B`, 64 is `BA`, and 77 is `BN`. We pad the left side with `A` (which stands for `0`) as needed: `AAABN`, and we remove the last character since we want to only change the path every 64 features, not every feature, giving us `AAAB`. (Feature filenames already have their own scheme which distiguishes them from every other feature in the same folder). Treating this as a path 4 levels long gives us `A/A/A/B`.
+
+So, feature with primary key values `[77]` would be written at `A/A/A/B/kU0=` using this path-structure.
```

Comment:
worth an example where we're encoding a much larger number to illustrate how we max out at 64^5 dirs and we use the LSBs to allocate to trees?

Reply:
Done